code/modules/mob/living/basic/slime/slime.dm

```diff
@@ -0,0 +1,385 @@
+	set_life_stage(new_life_stage)
+	update_name()
+	regenerate_icons()
+
+	set_nutrition(SLIME_STARTING_NUTRITION)
+
+	AddComponent(/datum/component/buckle_mob_effect,  mob_effect_callback = CALLBACK(src, PROC_REF(feed_process)))
+	AddComponent(/datum/component/health_scaling_effects, min_health_slowdown = 2)
+	AddComponent(/datum/component/obeys_commands, pet_commands)
+
+	AddElement(/datum/element/ai_retaliate)
+	AddElement(/datum/element/basic_health_examine, light_damage_message = "It has some punctures in its flesh!", heavy_damage_message = span_bold("It has severe punctures and tears in its flesh!"), heavy_threshold = 0.4)
+	AddElement(/datum/element/footstep, footstep_type = FOOTSTEP_MOB_SLIME)
+	AddElement(/datum/element/soft_landing)
+	AddElement(/datum/element/swabable, CELL_LINE_TABLE_SLIME, CELL_VIRUS_TABLE_GENERIC_MOB, 1, 5)
+
+	ADD_TRAIT(src, TRAIT_CANT_RIDE, INNATE_TRAIT)
+	ADD_TRAIT(src, TRAIT_VENTCRAWLER_ALWAYS, INNATE_TRAIT)
+
+	RegisterSignal(src, COMSIG_HOSTILE_PRE_ATTACKINGTARGET, PROC_REF(on_slime_pre_attack))
+	RegisterSignal(src, COMSIG_ATOM_ATTACK_HAND, PROC_REF(on_attack_hand) )
+
+	ai_controller.set_blackboard_key(BB_SLIME_FACE, SLIME_MOOD_NONE)
+	ai_controller.set_blackboard_key(BB_SLIME_EVOLVE, evolve_action)
+	ai_controller.set_blackboard_key(BB_SLIME_REPRODUCE, reproduce_action)
+
+/mob/living/basic/slime/Destroy()
+
+	UnregisterSignal(src, list(COMSIG_HOSTILE_PRE_ATTACKINGTARGET, COMSIG_ATOM_ATTACK_HAND))
+
+	QDEL_NULL(evolve_action)
+	QDEL_NULL(reproduce_action)
+
+	return ..()
+
+///Random slime subtype
+/mob/living/basic/slime/random/Initialize(mapload, new_colour, new_life_stage)
+	. = ..(mapload, pick(subtypesof(/datum/slime_type)), prob(50) ? SLIME_LIFE_STAGE_ADULT : SLIME_LIFE_STAGE_BABY)
+
+///Friendly docile subtype
+/mob/living/basic/slime/pet
+	hunger_disabled = TRUE
+
+/mob/living/basic/slime/pet/Initialize(mapload, new_colour, new_life_stage)
+	. = ..()
+	set_pacified_behaviour()
+
+//Hilbert subtype
+/mob/living/basic/slime/hilbert/Initialize(mapload, new_colour, new_life_stage)
+	. = ..(mapload, /datum/slime_type/bluespace)
+	ai_controller?.set_blackboard_key(BB_SLIME_RABID, TRUE)
+
+//Overriden base procs
+
+/mob/living/basic/slime/adjust_nutrition(change)
+	..()
+	nutrition = min(nutrition, SLIME_MAX_NUTRITION)
+
+/mob/living/basic/slime/update_name()
+	///Checks if the slime has a generic name, in the format of baby/adult slime (123)
+	var/static/regex/slime_name_regex = new("\\w+ (baby|adult) slime \\(\\d+\\)")
+	if(slime_name_regex.Find(name))
+		var/slime_id = rand(1, 1000)
+		name = "[slime_type.colour] [life_stage] slime ([slime_id])"
+		real_name = name
+	return ..()
+
+/mob/living/basic/slime/regenerate_icons()
+	cut_overlays()
+	var/icon_text = "[slime_type.colour] [life_stage] slime"
+	icon_dead = "[icon_text] dead"
+	if(stat != DEAD)
+		icon_state = icon_text
+		if(ai_controller?.blackboard[BB_SLIME_FACE] && !stat)
+			add_overlay("aslime-[ai_controller?.blackboard[BB_SLIME_FACE]]")
```

Comment:
How do player controlled slimes set a slime face? Or do player slimes have an AI controller that is innately disabled. (I guess to that end, in what contexts are ai_controller `null`?

Reply:
All mobs with an AI controller defined has an AI with its status turned off. Thus looks like it, the AI controller is actually never null, thus players will be always able to set these blackboard values via the emotes that set them. I can remove the ? checks.

Reply:
It looks like this is used all over the place in the codebase. Sometimes AI blackboard accessing in mob procs have the coalesce checks, and sometimes they do not.

Reply:
I have put the current mood back to a variable on the slimes. This way if the AI_controller is null for any reason, player controlled slimes will be able to still change their faces.